xdsl/dialects/vector.py

```diff
@@ -292,6 +295,104 @@ def get(mask_operands: list[Operation | SSAValue]) -> CreatemaskOp:
         )
 
 
+@irdl_op_definition
+class ExtractElementOp(IRDLOperation):
+    name = "vector.extractelement"
+    vector = operand_def(VectorType)
+    position = opt_operand_def(IndexTypeConstr | SignlessIntegerConstraint)
+    result = result_def(Attribute)
+    traits = traits_def(Pure())
+
+    def verify_(self):
+        assert isa(self.vector.type, VectorType[Attribute])
+
+        if self.result.type != self.vector.type.element_type:
+            raise VerifyException(
+                "Expected result type to match element type of vector operand."
+            )
+
+        if self.vector.type.get_num_dims() == 0:
+            if self.position is not None:
+                raise VerifyException("Expected position to be empty with 0-D vector.")
+            return
+        if self.vector.type.get_num_dims() != 1:
+            raise VerifyException("Unexpected >1 vector rank.")
+        if self.position is None:
+            raise VerifyException("Expected position for 1-D vector.")
+
+    def __init__(
+        self,
+        vector: SSAValue | Operation,
+        position: SSAValue | Operation | None = None,
+    ):
+        vector = SSAValue.get(vector)
+        assert isa(vector.type, VectorType[Attribute])
+
+        # Fail fast or rely on verification?
+        assert (vector.type.get_num_dims() == 1 and position is not None) or (
+            vector.type.get_num_dims() == 0 and position is None
+        )
```

Comment:
When in doubt, rely on verification, we have much better mechanisms to deal with that, and communicate it to the user, otherwise it's just an assert in the middle of a compilation pipeline without a nice error message.